### PR TITLE
[SPARK-42622][CORE] Disable substitution in values

### PR DESCRIPTION
**Pull request by jelmerk**
https://issues.apache.org/jira/browse/SPARK-42622

### What changes were proposed in this pull request?

Disable substitution in values for the `StringSubstitutor` used to resolve error messages

### Why are the changes needed?

when constructing an error message `ErrorClasssesJSONReader`

1. reads the error message from core/src/main/resources/error/error-classes.json 
2. replaces `<fieldValue>` with `${fieldValue}` in the error message it read
3. uses `StringSubstitutor` to replace `${fieldValue}` with the actual value

If `fieldValue` is defined as `"${foo}"` then it will also try and resolve foo. 
When foo is undefined it will throw an IllegalArgumentException

This is very problematic instance in this scenario. Where parsing json will fail if it does not adhere to the correct schema

image

Here the error message produced will be something like 

`Cannot parse the field name properties and the value ${foo} of the JSON token type string to target Spark data type struct.`

And because foo is undefined another error will be triggered, and another, and another.. until you have a stackoverflow.
It could be employed as a denial of service attack on data pipelines

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

Locally

Before

image

After

image

### CODE REVIEW

**Comment by sandeep-katta0102** [comment]
@jelmerk could you please raise a bug [here](https://issues.apache.org/jira/projects/SPARK/summary) and use this bug id in the PR for e.g. you can refer the PR https://github.com/apache/spark/pull/40202

**Comment by jelmerk** [comment]
> When would the field value have a substitution string in it - does this actually happen in practice?

I mention one such scenario in my description. Starting from the image on

**Comment by srowen** [comment]
Sorry I guess I mean I am not clear how this class, which reads error messages, is coming into play here. This is my ignorance. Seems like it wouldn't be related to just parsing user JSON, but the change fixes it?

**Comment by jelmerk** [comment]
> Sorry I guess I mean I am not clear how this class, which reads error messages, is coming into play here. This is my ignorance. Seems like it wouldn't be related to just parsing user JSON, but the change fixes it?

The example defines `properties` in the schema as a struct but in the json it is a string with the value "${foo}"
An exception is raised during parsing and the message is resolved by that class

**Comment by srowen** [comment]
Ah, I get it now. I agree with this change for sure.

**Comment by srowen** [comment]
(Can you just file a JIRA and link it per https://spark.apache.org/contributing.html ?)

**Comment by jelmerk** [comment]
> (Can you just file a JIRA and link it per https://spark.apache.org/contributing.html ?)

Already did that a few mins ago and added it on top of the PR description

**Comment by srowen** [comment]
Edit the title to link it. See other PRs. Minor thing but helps

**Comment by jelmerk** [comment]
> Edit the title to link it. See other PRs. Minor thing but helps

done

**Comment by srowen** [comment]
@dongjoon-hyun I'd like a second opinion - should we merge to 3.4 or 3.3? Reasonable fix I think and I'm a little concerned there could be some weird way this would expand some info it shouldn't

**Comment by dongjoon-hyun** [comment]
For branch-3.4, I'm +1 for backporting.
Does this affect branch-3.3 too, @srowen ?

**Comment by dongjoon-hyun** [comment]
Since [SPARK-40530](https://issues.apache.org/jira/browse/SPARK-40530) added this, `branch-3.3` seems to be not affected.

**Comment by dongjoon-hyun** [comment]
Also, cc @cloud-fan , @MaxGekk , @viirya from SPARK-40530.

File: core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala [hunk 1]
```diff
@@ -47,6 +47,7 @@ class ErrorClassesJsonReader(jsonFileURLs: Seq[URL]) {
     val messageTemplate = getMessageTemplate(errorClass)
     val sub = new StringSubstitutor(messageParameters.asJava)
     sub.setEnableUndefinedVariableException(true)
+    sub.setDisableSubstitutionInValues(true)
```
Comment (wangyum) [inline core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala:50]: Could we add a test case?

**Comment by srowen** [comment]
Oh yeah, won't affect 3.3. A quick test like your proof of concept would be great indeed.

File: core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala (hunk 1 above)
Comment (jelmerk) [reply core/src/main/scala/org/apache/spark/ErrorClassesJSONReader.scala:50]: I added a test in SparkThrowableSuite

**Comment by jelmerk** [comment]
> Oh yeah, won't affect 3.3. A quick test like your proof of concept would be great indeed.

The class itself is not present on 3.3. It is not affected.  In-fact thats how we mitigated this problem, that occurred in a production setting for our pipeline. by downgrading

**Comment by dongjoon-hyun** [comment]
Wow. Thank you for sharing the details, @jelmerk .

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
good catch!

**Review by wangyum — APPROVED** [review]

**Comment by viirya** [comment]
The fix looks okay, but the PR description looks confusing as `What changes were proposed in this pull request?` describes an alternative approach to current `getErrorMessage`?

**Review by viirya — COMMENTED** [review]
Please make sure updating the description before merging this.

**Comment by jelmerk** [comment]
> Please make sure updating the description before merging this.

I removed the alternative solution. Imho It's a much better solution, but if experience has taught me anything getting anything but the smallest of changes merged is an uphill battle so I chose the 1 line fix

**Review by MaxGekk — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by jelmerk** [comment]
> Simple is good. Indeed the smallest of changes can even break things. This looks good pending CI/CD tests

I tried triggering it twice but those tests seem unstable

**Comment by viirya** [comment]
Can you re-trigger it again? Although the CI failure looks unrelated.

**Comment by srowen** [comment]
Hm, no looks like something else is wrong in even setting up the environment. Github issue?
`Error: Error response from daemon: Get "https://ghcr.io/v2/": received unexpected HTTP status: 503 Service Unavailable`

**Comment by srowen** [comment]
Merged to master/3.4

**Pull request closed without merging**